misk-core/src/main/kotlin/misk/sampling/RateLimiter.kt

```diff
@@ -107,6 +107,45 @@ class RateLimiter private constructor(
     }
   }
 
+  fun getPermitsRemaining(
+    unit: TimeUnit,
+    permitCount: Long,
+    timeout: Long
+  ): Long {
+    while (true) {
+      val allocatedUntil = atomicAllocatedUntil.get()
+      val permitsPerSecond = this.permitsPerSecond // Sample this volatile only once.
+
+      val maxRequestSize = windowSizeNs.nanosToPermits(permitsPerSecond)
+      if (permitCount > maxRequestSize) return 0L
+
+      val now = ticker.read()
+
+      // If this acquire succeeds, this is the time we're consuming permits through.
+      val newAllocatedUntil = maxOf(allocatedUntil, now) +
+        permitCount.permitsToNanos(permitsPerSecond)
+      println(newAllocatedUntil)
+
+      val sleepNs = newAllocatedUntil - now - windowSizeNs
+      println("windowSizeNs $windowSizeNs")
+      println("sleepNs $sleepNs")
+
+      val timeoutNs = unit.toNanos(timeout)
+
+      // We'd have to sleep too long for this number of permits. Fail fast.
+      if (sleepNs > timeoutNs) return 0L
+
+      // Try to consume permits! If atomicAllocatedUntil changed, we lost a race; loop to try again.
+      if (!atomicAllocatedUntil.compareAndSet(allocatedUntil, newAllocatedUntil)) continue
```

Comment:
When I saw the method name `getPermitsRemaining` I assumed that it just returned how many permits were remaining, but it looks like this method actually acquires permits (like `tryAcquire`) and then returns how many permits were remaining after acquiring. In other words, it looks like this method is a reimplementation of `tryAcquire`, but it just returns different information.

I think we should do one of two things:

1. Rename this method to something like `tryAcquire` (though we can't use that name exactly because it would conflict), and possibly deprecate the existing method.
2. Trim down this method significantly so that all it does is get how many permits are remaining, and has no side effects at all, including acquiring any permits.

The second option feels simpler. I imagine these methods be used by the caller like:

```
if(!rateLimiter.tryAcquire(...)) {
  throw RuntimeException("too busy, go away")
}

val permitsRemaining = rateLimiter.getPermitsRemaining()
metricRegistry.recordGauge("permits-remaining-percent", permitsRemaining/maxPermits)
```

On one hand, this would be subject to some race conditions, because two threads might take out two permits at the same time, and then both threads would record that there is `X` percent remaining. On the other hand, I don't think it really matters if this number is subject to race conditions; all anyone will typically care about is the `min` value that occurs in a given time period.

tl;dr I recommend chopping this method way down so that all it does is calculate the `permitsRemaining` and returns that.